docs/design_proposals/new-render-and-deploy.md

```diff
@@ -0,0 +1,374 @@
+| `skaffold deploy` |`deploy.dir` not given|*Render* and deploy
+`kpt pipeline run --sink-mode=stdout | kpt live apply --`|
+| `skaffold deploy` |`deploy.dir: [PATH]` |Only deploy `kpt live apply [PATH]`|
+|`skaffold deploy --render-path [PATH2]`| `deploy.dir: [PATH1]`|Deploy and override PATH1 `kpt live apply [PATH2]`|
+| `skaffold dev` | | Build, render and deploy|
+
+### Integration of `kpt` pipeline and `skaffold` render 
+
+Either *kpt-managed* Pipeline or *skaffold-managed* Pipeline is driven by a `kpt fn render` command (see below). This requires a `Kptfile` file. The *kpt-managed* approach finds the `Kptfile` from `.render.kpt`, and the *skaffold-managed* approach creates the `Kptfile` file on the fly.
+
+![render command](images/render-cmd.png)
+
+This `Kptfile` is built as shown below. Since kpt considers all resources under the `Kptfile` directory as the source. `skaffold` will create a temporary directory `.kpt-hydrated` to cache the `kustomize` DRY configs, `helm` charts, and the raw manifests. Before running `kpt`, these resources are copied to the temporary directory, and `kpt` will hydrate the configurations in place.
+
+``` yaml
+apiVersion: kpt.dev/v1alpha1
+kind: Kptfile
+metadata:
+  name: skaffold-kpt-builtin-pipeline
+pipeline:
+   mutators:
+
+      # items directly defined in skaffold.yaml. It can only be the tools skaffold whitelisted. 
+      - image: <render.transform.customize[0].name> 
+      # specific arguments that could defer per whitelisted tool.
+        configPath/configPath:      
+
+      - image: <render.transform.customize[1].name> 
+      ...
+
+
+   validators:
+  
+      # items directly defined in skaffold.yaml. It can only be the tools skaffold whitelisted. E.g. kubeval. 
+      - image: <render.validators.customize[0].name> 
+      # specific arguments that could defer per whitelisted tool.
+        configPath/configPath: 
+
+      - image: <render.validators.customize[1].name> 
+      ...
+
+``` 
+
+The following graph describes the commands and operations to generate the kpt sources.
+
+![builtin render](images/builtin-render.png)
+
+### Integration of `kpt` live  and `skaffold` deploy
+
+`kpt` relies on `ResourceGroup` to accurately deploy the WET configurations to the cluster.  This requires the ResourceGroup CRD to be installed on the cluster side. 
+
+`skaffold dev` and `skaffold deploy` commands will detect and ask for user permissions before installing the `ResourceGroup` controller. It calls the following `kpt` commands. 
+
+![deploy command](images/deploy-cmd.png)
+
+*Note*: `skaffold` v2 will not provide the migration from `kpt` v0.X to `kpt` v1. Users can follow [this guide](https://googlecontainertools.github.io/kpt/reference/live/migrate/) to establish the `kpt` migration themselves.
+
+### Builtin `kpt`
+
+`skaffold` will have `kpt` embedded via go bindata, which ship together with the skaffold binary to guarantee the versions match.
```

Comment:
Is there any performance concern w/ this approach?  Also it might make sense to add a blurb around how this will be tested specifically as I don't believe skaffold has integration deps that pull in external deps atm

Reply:
Also it might make sense to add a blurb around how this will be tested specifically as I don't believe skaffold has integration tests that pull in external deps atm.  Will skaffold pin to a kpt version and be updated ad-hoc?  Use the latest kpt release?  Will skaffold test this then at release time.